### PR TITLE
Product detail page part 1 

### PR DESCRIPTION
**Pull request by zeina-idris**
## Checklist
* [x] Unit tests
* [x] E2E tests
* [ ] Documentation
____________________________
* Implement product detail page
* Add additional e2e tests to product thumbnail for enabling link to PDP
* Add currency computed property in priceMixin file 
____________________________
## Other points of interest
* Use **vue product zoomer** library for product gallery
* Use **variant** apollo field for attributes by sku
* Move jQuery toggle method to product info component

### CODE REVIEW

File: src/components/ProductInfo.vue [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+      <ProductGallery v-if="images"
+                      :productImages="images" />
+    </div>
+    <div data-test="product-description"
+         class="col-sm-6 product-description">
+      <div class="row">
+        <div class="col-sm-12">
+            <h1 data-test="pdp-product-title"
+                class="text-uppercase pdp-product-title">
+              {{currentProduct.name}}
```
Comment (lauraluiz) [inline src/components/ProductInfo.vue]: Unfortunately ESLint doesn't complain about it, but usually we leave whitespaces around the variable name: `{{ currentProduct.name }}`.

File: src/components/ProductInfo.vue [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+      </div>
+      <!-- {{> catalog/add-to-cart}}
+      {{> catalog/add-to-wishlist-btn}}
+      {{> catalog/reserve-in-store-btn}} -->
+      <button id=""
+              data-test="add-to-cart-button"
+              class="add-to-bag-btn">
+        <img class="bag-thumb"
+             src="../assets/img/hand-bag-2-black.png"
+             alt="$t('catalog.cart.add')">
```
Comment (lauraluiz) [inline src/components/ProductInfo.vue]: This should also be moved into the local `<i18n>`, right? If it's used anywhere else what we can do is create a component for this button and put it there.

File: src/components/ProductInfo.vue [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+export default {
+  components: {
+    ProductGallery,
+  },
+
+  data: () => ({
+    product: {},
+  }),
+
+  props: {
```
Comment (lauraluiz) [inline src/components/ProductInfo.vue]: I would put props before data because it's more important (at least it's more relevant to people using this component), but I'm not sure how we did it in other components...

File: src/components/ProductInfo.vue [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+                      designer {
+                        label
+                        key
+                        name
+                      }
+                      colorFreeDefinition {
+                        value(locale: $locale)
+                        name
+                      }
+                      size{
```
Comment (lauraluiz) [inline src/components/ProductInfo.vue]: A whitespace between `size` and `{` 😬

File: src/components/ProductInfo.vue [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+        }
+
+        fragment printPrice on BaseMoney {
+          centAmount
+          fractionDigits
+        }`,
+      variables() {
+        return {
+          locale: this.$i18n.locale,
+          currency: 'EUR',
```
Comment (lauraluiz) [inline src/components/ProductInfo.vue]: This shouldn't be hardcoded.

File: src/components/ProductThumbnail.vue [hunk 6]
```diff
@@ -147,6 +147,14 @@ export default {
     originalPrice() {
       return this.matchingVariant.price.value;
     },
+
+    productSlug() {
+      return this.currentProduct.slug;
+    },
+
+    sku() {
+      return this.matchingVariant.sku;
```
Comment (lauraluiz) [inline src/components/ProductThumbnail.vue]: I'm not sure it's always the same structure, but any method that you see repeated across components we should move it into its own mixin to make sure that we never have repeated code. For example, `currentProduct` I think it appears several times, right?

File: tests/e2e/specs/productDetailPage_spec.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    cy.visit('/product/lemare-booties-0778-grey/M0E20000000E0WX');
+  });
+
+  it('Displays a single product details', () => {
+    cy.get('[data-test=pdp-product-title]', { timeout: 20000 })
+      .contains('Booties Lemare grey')
+      .parentsUntil('[data-test=product-description]')
+      .parent()
+      .find('[data-test=quickview-sku]')
+      .contains('M0E20000000E0WX')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: If you are testing here whether the SKU is the correct one, then this should be a `should` statement, probably inside the `.then(($product) => ...)`.

File: tests/e2e/specs/productDetailPage_spec.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  before(() => {
+    cy.visit('/product/lemare-booties-0778-grey/M0E20000000E0WX');
+  });
+
+  it('Displays a single product details', () => {
+    cy.get('[data-test=pdp-product-title]', { timeout: 20000 })
+      .contains('Booties Lemare grey')
+      .parentsUntil('[data-test=product-description]')
+      .parent()
+      .find('[data-test=quickview-sku]')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: If we are in the PDP there shouldn't be any quickview, the quickview is that summary of the product that appears in the POP. Use a more descriptive name.

File: tests/e2e/specs/productDetailPage_spec.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      .contains('M0E20000000E0WX')
+      .parentsUntil('[data-test=product-description]')
+      .parent()
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: There are lots of moving up and down the DOM. Sometimes it's unavoidable (unfortunately) but I hope this isn't one of those cases. It would be easier to understand and maintain in the future if the test consisted of:
1) find the place where the info is
2) assert each of the info is correct
3) repeat until all the info is checked
If the path to find the place where the info is, is too complex, then it can be grouped under a `.then(($someInfo) => ...)` to avoid duplication of code.

So in short: try to improve this code 😄

File: tests/e2e/specs/productDetailPage_spec.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        cy.wrap($product)
+          .find('[data-test=product-discounted-price]')
+          .should('contain', '248,75 €');
+        cy.wrap($product)
+          .find('[data-test=product-discount-price]')
+          .should('contain', '174,12 €');
+
+        cy.wrap($product)
+          .find('[data-test=add-to-cart-button]')
+          .should('exist');
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: This, as it is, it's a bit unnecessary. We will better test this button on the add to cart e2e.

File: tests/unit/specs/components/ProductGallery.spec.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        },
+        {
+          id: 1,
+          url: 'https://s3-eu-west-1.amazonaws.com/commercetools-maximilian/products/079535_1_large.jpg',
+        },
+      ],
+    };
+    const wrapper = shallowMount(ProductGallery, options);
+    wrapper.setProps({ productImages });
+    expect(wrapper.vm.images).toEqual(images);
```
Comment (lauraluiz) [inline tests/unit/specs/components/ProductGallery.spec.js]: This is tiny but I think it helps readability. Instead of declaring the `const images` at the beginning of the test and then referring to it in this last assertion, better declare it directly in the assertion:
```
expect(wrapper.vm.images).toEqual({
  thumbs: [...],
  normal_size: [...],
  large_size: [...]
});
```

File: src/components/ProductInfo.vue (hunk 5 above)
Comment (zeina-idris) [reply src/components/ProductInfo.vue]: I temporarily copied number formats object from i18n file locally to price mixin file so we can use it until we find a solution for the issue after I create it.

File: src/components/ProductInfo.vue [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+              </span>
+            </span>
+          </p>
+
+        </div>
+      </div>
+      <!-- {{> catalog/add-to-cart}}
+      {{> catalog/add-to-wishlist-btn}}
+      {{> catalog/reserve-in-store-btn}} -->
+      <button id=""
```
Comment (lauraluiz) [inline src/components/ProductInfo.vue]: If it's going to be empty, you should better remove it.

File: tests/e2e/specs/productDetailPage_spec.js [hunk 13]
```diff
@@ -0,0 +1,46 @@
+describe('Product detail page', () => {
+  before(() => {
+    cy.visit('/product/lemare-booties-0778-grey/M0E20000000E0WX');
+  });
+
+  it('Displays a single product details', () => {
+    cy.get('[data-test=product-description]', { timeout: 20000 })
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: Since "product description" is actually a field of a product, I would use a different name to refer to the product data (i.e. name, description, attributes, price, etc.). Maybe "product-data", or "product".

File: tests/e2e/specs/productDetailPage_spec.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+      .then(($product) => {
+        cy.wrap($product)
+          .find('[data-test=pdp-product-title]')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: Why "pdp-product-title" and not simply "product-name"?

File: tests/e2e/specs/productDetailPage_spec.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+        cy.wrap($product)
+          .find('[data-test=product-sku]')
+          .should('contain', 'M0E20000000E0WX');
+
+        cy.wrap($product)
+          .find('[data-test=product-discounted-price]')
+          .should('contain', '248,75 €');
+        cy.wrap($product)
+          .find('[data-test=product-discount-price]')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: "product-discounted-price" vs "product-discount-price". I'm not sure of the difference.

File: tests/e2e/specs/productDetailPage_spec.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+          .should('contain', '174,12 €');
+
+        cy.wrap($product)
+          .find('[data-test=panel-default]')
+          .should('exist')
+          .find('[data-test=accordion-toggle]')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: For the `data-test`, try to write what you feel that the `div` contains, don't base your name on the CSS styles that you see, because they are generally bad.
So given that, I hope you would consider naming these both different than `panel-default` and `accordion-toggle`.

File: tests/e2e/specs/productDetailPage_spec.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+          .click();
+
+        cy.wrap($product)
+          .find('[data-test=list-attributes]')
+          .should('have.length', 7)
+          .eq(2)
+          .should('contain', 'size:')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: I guess you can replace `contain` by `eq`, since it is exactly what you wrote there and not a substring.

File: tests/e2e/specs/productDetailPage_spec.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+          .parent()
+          .should('contain', '34');
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: Same here, regarding `eq`

File: tests/e2e/specs/productDetailPage_spec.js [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+      });
+
+    cy.get('[data-test=product-gallery]')
+      .find('[data-test=product-image]')
+      .should('exist');
+
+    cy.get('.thumb-list > .responsive-image')
```
Comment (lauraluiz) [inline tests/e2e/specs/productDetailPage_spec.js]: CSS? D:

File: tests/unit/specs/components/ProductThumbnail.spec.js [hunk 20]
```diff
@@ -1,23 +1,7 @@
...
-  const discountedPrice = {
-    discounted: {
-      value: {
-        centAmount: 1000,
-        fractionDigits: 2,
-      },
-    },
-  };
-
+describe.skip('ProductThumbnail.vue', () => {
```
Comment (lauraluiz) [inline tests/unit/specs/components/ProductThumbnail.spec.js]: Remove the `skip`

File: tests/unit/specs/components/ProductInfo.spec.js [hunk 21]
```diff
@@ -0,0 +1,59 @@
+import { shallowMount } from '@vue/test-utils';
+import ProductInfo from '@/components/ProductInfo.vue';
+
+describe.skip('ProductInfo.vue', () => {
```
Comment (lauraluiz) [inline tests/unit/specs/components/ProductInfo.spec.js]: Remove the `skip`
